luaui/Widgets/cmd_blueprint.lua

```diff
@@ -272,6 +274,70 @@ local function setSelectedBlueprintIndex(index)
 	end
 end
 
+local function isValidBlueprint(blueprint)
+	if blueprint == nil then
+		return false
+	end
+
+	if blueprint.hasInvalidUnits then
+		return false
+	end
+
+	local buildable, unbuildable = WG["api_blueprint"].getBuildableUnits(blueprint)
+
+	if buildable == 0 then
+		return false
+	end
+
+	return true
+end
+
+local function getNextFilteredBlueprintIndex(startIndex, checkStart)
+	local newIndex = startIndex or selectedBlueprintIndex or 1
+	local foundValid = isValidBlueprint(blueprints[newIndex])
+
+	if foundValid and checkStart then
+		return newIndex
+	end
+
+	for _ = 1, #blueprints do
+		newIndex = nextIndex(newIndex, #blueprints)
+		if isValidBlueprint(blueprints[newIndex]) then
+			foundValid = true
+			break
+		end
+	end
+
+	if foundValid then
+		return newIndex
+	else
+		return nil
+	end
+end
+
+local function getPrevFilteredBlueprintIndex(startIndex, checkStart)
+	local newIndex = startIndex or selectedBlueprintIndex or 1
+	local foundValid = isValidBlueprint(blueprints[newIndex])
+
+	if foundValid and checkStart then
+		return newIndex
+	end
+
+	for _ = 1, #blueprints do
+		newIndex = prevIndex(newIndex, #blueprints)
+		if isValidBlueprint(blueprints[newIndex]) then
+			foundValid = true
```

Comment:
I think you can just return newIndex here and avoid break and the tests for foundValid at the end of the function.

Reply:
sounds reasonable